# Phase 1 — Data Preprocessing
## Author: Rashika Agarwal
## Objective

The purpose of this notebook is to prepare Apple's historical stock market data for exploratory analysis and predictive modelling.

The preprocessing workflow includes:

- Loading the raw dataset
- Inspecting the dataset structure
- Examining descriptive statistics
- Checking for missing values
- Identifying duplicate records
- Validating data types and date indexing
- Organizing the dataset chronologically
- Saving a cleaned version for subsequent analysis

The cleaned dataset generated in this notebook will serve as the input for all later stages of the project, including exploratory data analysis, feature engineering, machine learning, and forecasting.

# 1. Import Required Libraries

The required Python libraries are imported for data manipulation and preprocessing.

In [4]:
import pandas as pd
import numpy as np

# 2. Load Raw Apple Stock Dataset

The raw Apple stock dataset is loaded from the project's data directory. The Date column is parsed as a datetime object and used as the DataFrame index to facilitate time-series analysis.

In [8]:
apple = pd.read_csv(
    "../data/raw/AAPL.csv",
    index_col="Date",
    parse_dates=True
)

# 3. Preview the Dataset

The first five observations are displayed to verify that the dataset has been loaded correctly and to gain an initial understanding of its structure.

In [9]:
apple.head()

,Close,High,Low,Open,Volume
Date,,,,,
2020-01-02,72.333885,72.394093,71.091191,71.344062,135480400
2020-01-03,71.630661,72.389280,71.406689,71.563228,146322800
2020-01-06,72.201431,72.239965,70.503569,70.754036,118387200
2020-01-07,71.861839,72.466322,71.642681,72.211041,108872000
2020-01-08,73.017853,73.318893,71.565636,71.565636,132079200


# 4. Dataset Structure

The dataset structure is examined to determine:

- Number of observations
- Number of variables
- Data types
- Missing values
- Memory usage

In [10]:
apple.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1628 entries, 2020-01-02 to 2026-06-25
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Close   1628 non-null   float64
 1   High    1628 non-null   float64
 2   Low     1628 non-null   float64
 3   Open    1628 non-null   float64
 4   Volume  1628 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 76.3 KB


### Key Insights

- The dataset contains **1,628 trading days** spanning from **2 January 2020 to 25 June 2026**.
- Five numerical variables are available: Close, High, Low, Open, and Volume.
- All columns contain complete observations with no missing values.
- The Date column has been successfully stored as a DatetimeIndex, making the dataset suitable for time-series analysis.

# 5. Statistical Summary

Descriptive statistics provide an overview of the central tendency, spread, and overall distribution of each numerical feature.

In [11]:
apple.describe()

,Close,High,Low,Open,Volume
count,1628.000000,1628.000000,1628.000000,1628.000000,1.628000e+03
mean,172.603195,174.390870,170.683961,172.462315,8.188450e+07
std,55.737273,56.179436,55.327491,55.738475,5.049625e+07
min,54.163704,55.160702,51.324800,55.059313,1.791060e+07
25%,134.593174,136.302051,132.415615,134.507080,4.890652e+07
50%,168.251762,169.689272,166.719160,167.915708,6.793505e+07
75%,213.077183,215.308501,211.005907,213.064585,9.690462e+07
max,315.200012,317.399994,309.649994,314.179993,4.265100e+08


### Key Insights

- Apple's average closing price during the study period is approximately **172 USD**.
- The stock reached a maximum closing price of approximately **315 USD**, reflecting substantial long-term growth.
- Trading volume varies considerably, indicating periods of elevated investor activity.
- The wide range between minimum and maximum prices suggests significant market appreciation over the analysis period.

# 6. Missing Value Analysis

Missing values are identified to ensure the dataset is complete before analysis.

In [12]:
apple.isnull().sum()

Close     0
High      0
Low       0
Open      0
Volume    0
dtype: int64

### Key Insights

- No missing values were identified in any column.
- No imputation or data cleaning is required before analysis.

# 7. Duplicate Record Analysis

Duplicate observations are checked to prevent redundant information from affecting future analyses.

In [21]:
duplicates = apple.duplicated().sum()

print(f"Duplicate rows: {duplicates}")

Duplicate rows: 0


### Key Insights

- No duplicate rows were detected.
- Each trading day appears exactly once in the dataset.

# 8. Dataset Dimensions

The overall size of the dataset is examined.

In [22]:
rows, columns = apple.shape

print(f"Rows    : {rows}")
print(f"Columns : {columns}")

Rows    : 1628
Columns : 5


### Key Insights

- The dataset contains **1,628 observations**.
- Five numerical variables are available for analysis.

# 9. Column Names

The available variables are displayed.

In [23]:
list(apple.columns)

['Close', 'High', 'Low', 'Open', 'Volume']

### Key Insights

The dataset contains the following variables:

- Open
- High
- Low
- Close
- Volume

These variables provide the essential market information required for financial analysis.

# 10. Date Index Inspection

The Date index is inspected to verify chronological ordering and datetime formatting.

In [16]:
apple.index

DatetimeIndex(['2020-01-02', '2020-01-03', '2020-01-06', '2020-01-07',
               '2020-01-08', '2020-01-09', '2020-01-10', '2020-01-13',
               '2020-01-14', '2020-01-15',
               ...
               '2026-06-11', '2026-06-12', '2026-06-15', '2026-06-16',
               '2026-06-17', '2026-06-18', '2026-06-22', '2026-06-23',
               '2026-06-24', '2026-06-25'],
              dtype='datetime64[ns]', name='Date', length=1628, freq=None)

### Key Insights

- The dataset is indexed using trading dates.
- The Date index is stored as a DatetimeIndex.
- This structure enables efficient time-series calculations such as rolling averages, resampling, and forecasting.

# 11. Validate Data Types

The data type of each variable is verified.

In [24]:
pd.DataFrame({
    "Column": apple.columns,
    "Data Type": apple.dtypes.values
})

,Column,Data Type
0,Close,float64
1,High,float64
2,Low,float64
3,Open,float64
4,Volume,int64


### Key Insights

- Price variables are stored as floating-point values.
- Trading volume is stored as an integer.
- The data types are appropriate for numerical analysis and machine learning.

# 12. Sort Dataset

The dataset is sorted chronologically to ensure the observations follow the correct trading sequence.

In [18]:
apple.sort_index(inplace=True)

apple.head()

,Close,High,Low,Open,Volume
Date,,,,,
2020-01-02,72.333885,72.394093,71.091191,71.344062,135480400
2020-01-03,71.630661,72.389280,71.406689,71.563228,146322800
2020-01-06,72.201431,72.239965,70.503569,70.754036,118387200
2020-01-07,71.861839,72.466322,71.642681,72.211041,108872000
2020-01-08,73.017853,73.318893,71.565636,71.565636,132079200


# 13. Save Clean Dataset

The cleaned dataset is saved to the processed data directory. This dataset will be used throughout the remainder of the project.

In [19]:
apple.reset_index().to_csv(
    "../data/processed/apple_clean.csv",
    index=False
)

In [20]:
pd.read_csv("../data/processed/apple_clean.csv").head()

,Date,Close,High,Low,Open,Volume
0,2020-01-02,72.333885,72.394093,71.091191,71.344062,135480400
1,2020-01-03,71.630661,72.389280,71.406689,71.563228,146322800
2,2020-01-06,72.201431,72.239965,70.503569,70.754036,118387200
3,2020-01-07,71.861839,72.466322,71.642681,72.211041,108872000
4,2020-01-08,73.017853,73.318893,71.565636,71.565636,132079200


# 14. Preprocessing Summary

The preprocessing stage has been completed successfully.

### Summary

- Raw Apple stock dataset loaded successfully.
- Dataset structure validated.
- No missing values detected.
- No duplicate observations identified.
- Data types verified.
- Trading dates validated and maintained as a chronological index.
- Clean dataset exported to the processed data directory.

The cleaned dataset is now ready for exploratory data analysis in the next notebook.